samples/testapp/src/androidMain/kotlin/com/android/identity/testapp/NdefService.kt

```diff
@@ -0,0 +1,217 @@
+package com.android.identity.testapp
+
+import android.content.Intent
+import android.nfc.cardemulation.HostApduService
+import android.os.Bundle
+import android.os.Vibrator
+import androidx.core.content.ContextCompat
+import com.android.identity.cbor.DataItem
+import com.android.identity.crypto.Crypto
+import com.android.identity.crypto.EcCurve
+import com.android.identity.crypto.EcPrivateKey
+import com.android.identity.mdoc.connectionmethod.ConnectionMethod
+import com.android.identity.mdoc.connectionmethod.ConnectionMethodBle
+import com.android.identity.mdoc.nfc.MdocNfcEngagementHelper
+import com.android.identity.testapp.presentation.MdocPresentationMechanism
+import com.android.identity.mdoc.transport.MdocTransport
+import com.android.identity.mdoc.transport.MdocTransportFactory
+import com.android.identity.mdoc.transport.MdocTransportOptions
+import com.android.identity.nfc.CommandApdu
+import com.android.identity.testapp.presentation.PresentationModel
+import com.android.identity.util.AndroidContexts
+import com.android.identity.util.Logger
+import com.android.identity.util.UUID
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.datetime.Clock
+import kotlinx.io.bytestring.ByteString
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+
+// TODO: whether to use static or negotiated handover + other settings should be
+//  configurable in IsoMdocProximitySharingScreen
+
+class NdefService: HostApduService() {
+    companion object {
+        private val TAG = "NdefService"
+
+        private var engagement: MdocNfcEngagementHelper? = null
+        private var listeningTransports = mutableListOf<MdocTransport>()
+        private var disableEngagementJob: Job? = null
```

Comment:
There are some race conditions around all of these variables, depending on what's calling the functions that access them. Add locks around them?

Reply:
I don't think that's possible - the OS won't create multiple simultaneous instances of this class.